src/porepy/numerics/ad/secondary_operator.py

```diff
@@ -0,0 +1,897 @@
+        if get_derivatives:
+            kwargs["name"] = self._name_diffs
+        else:
+            kwargs["name"] = self.name
+
+        kwargs["data"] = self._data_of(grid)
+
+        if op.is_previous_time:
+            kwargs["time_step_index"] = op.time_step_index
+        else:
+            kwargs["iterate_index"] = op.iterate_index
+
+        return pp.get_solution_values(**kwargs)
+
+    def _data_of(self, grid: pp.GridLike) -> dict:
+        """Convenience function to get the data dictionary of any grid."""
+        if isinstance(grid, pp.Grid):
+            data = self.mdg.subdomain_data(grid)
+        elif isinstance(grid, pp.MortarGrid):
+            data = self.mdg.interface_data(grid)
+        elif isinstance(grid, pp.BoundaryGrid):
+            data = self.mdg.boundary_grid_data(grid)
+        else:
+            raise ValueError(f"Unknown grid type {type(grid)}.")
+        return data
+
+    @property
+    def num_dependencies(self) -> int:
+        """Number of first order dependencies of this operator, passed at instantiation.
+
+        This determines the number of required derivative values."""
+        return len(self._dependencies)
+
+    def num_dofs_on_grid(self, grid: pp.GridLike) -> int:
+        """Computes the numbe of DOFs based on the information provided during
+        instantiation.
+
+        Note:
+            On boundary grids and interfaces, only cell-wise DOFs are expected.
+            On subdomains, DOFs can be cell-, node- and face-wise.
+
+        Parameters:
+            grid: Any grid in the md-grid.
+
+        Returns:
+            For boundary grids and interfaces it returns ``grid.num_cells`` multiplied
+            with ``dof_info['cells']``.
+            For subdomains, it returns information based on ``dof_info``,
+            number of cells, number of faces and number of nodes in ``grid``.
+
+        Raises:
+            TypeError: If ``grid`` is neither a boundary, interface or subdomain.
+
+        """
+        if isinstance(grid, (pp.BoundaryGrid, pp.MortarGrid)):
+            # NOTE using default value of 1, because this is the general default value
+            # and only cells are supported on boundaries and interfaces
+            return self._dof_info.get("cells", 1) * grid.num_cells
+        elif isinstance(grid, pp.Grid):
+            # NOTE cannot use default value of scalar, cell-wise, to not mess with
+            # cases where the user defines only node- or face-wise dofs.
+            return (
+                self._dof_info.get("cells", 0) * grid.num_cells
+                + self._dof_info.get("faces", 0) * grid.num_faces
+                + self._dof_info.get("nodes", 0) * grid.num_nodes
+            )
+        else:
+            raise TypeError(f"Unsupported type of grid {type(grid)}.")
+
+    @property
+    def name(self) -> str:
+        """Name of this expression assigned at instantiation."""
+        return self._name
+
+    @property
+    def _name_diffs(self) -> str:
```

Comment:
`_name_derivatives`? I found `_name_diffs` very confusing for some reason.

Reply:
Changed to _name_derivatives